conductor_api/src/keystore.rs

```diff
@@ -0,0 +1,763 @@
+            hash_config,
+        })
+    }
+
+    #[allow(dead_code)]
+    pub fn new_from_file(
+        path: PathBuf,
+        passphrase_manager: Arc<PassphraseManager>,
+        hash_config: Option<PwHashConfig>,
+    ) -> HcResult<Self> {
+        let mut file = File::open(path)?;
+        let mut contents = String::new();
+        file.read_to_string(&mut contents)?;
+        let mut keystore: Keystore = serde_json::from_str(&contents)?;
+        keystore.hash_config = hash_config;
+        keystore.passphrase_manager = Some(passphrase_manager);
+        Ok(keystore)
+    }
+
+    #[allow(dead_code)]
+    fn check_passphrase(&self, mut passphrase: &mut SecBuf) -> HcResult<bool> {
+        let mut decrypted_buf = decrypt_with_passphrase_buf(
+            &self.passphrase_check,
+            &mut passphrase,
+            self.hash_config.clone(),
+            PCHECK_SIZE,
+        )?;
+        let mut decrypted_header = SecBuf::with_insecure(PCHECK_HEADER_SIZE);
+        let decrypted_buf = decrypted_buf.read_lock();
+        decrypted_header.write(0, &decrypted_buf[0..PCHECK_HEADER_SIZE])?;
+        let mut expected_header = SecBuf::with_secure(PCHECK_HEADER_SIZE);
+        expected_header.write(0, &PCHECK_HEADER)?;
+        Ok(decrypted_header.compare(&mut expected_header) == 0)
+    }
+
+    #[allow(dead_code)]
+    pub fn change_passphrase(
+        &mut self,
+        old_passphrase: &mut SecBuf,
+        new_passphrase: &mut SecBuf,
+    ) -> HcResult<()> {
+        if !self.check_passphrase(old_passphrase)? {
+            return Err(HolochainError::ErrorGeneric("Bad passphrase".to_string()));
+        }
+        self.passphrase_check = make_passphrase_check(new_passphrase, self.hash_config.clone())?;
+        Ok(())
+    }
+
+    fn decrypt(&mut self, id_str: &String) -> HcResult<()> {
+        let blob = self
+            .secrets
+            .get(id_str)
+            .ok_or(HolochainError::new("Secret not found"))?;
+        let mut passphrase = self.passphrase_manager.as_ref()?.get_passphrase()?;
+        let secret = match blob.blob_type {
+            BlobType::Seed => {
+                Secret::Seed(Seed::from_blob(blob, &mut passphrase, self.hash_config.clone())?.buf)
+            }
+            BlobType::SigningKey => Secret::SigningKey(SigningKeyPair::from_blob(
+                blob,
+                &mut passphrase,
+                self.hash_config.clone(),
+            )?),
+            BlobType::EncryptingKey => Secret::EncryptingKey(EncryptingKeyPair::from_blob(
+                blob,
+                &mut passphrase,
+                self.hash_config.clone(),
+            )?),
+            _ => {
+                return Err(HolochainError::ErrorGeneric(format!(
+                    "Tried to decrypt unsupported BlobType in Keystore: {}",
+                    id_str
+                )));
+            }
+        };
+        self.cache
+            .insert(id_str.clone(), Arc::new(Mutex::new(secret)));
```

Comment:
I'm uncertain about having a cache here. Part of the point of the passphrase manager is to give users control over "locking" their authorizor capabilities after an expiration timer, but if they authorize something and their authorizor key gets cached here, then we have to coordinate expiring the cache when the passphrase expires?

Reply:
Hm, I get that.

But we do need to hold some secrets unencrypted indefinitely: agent keys in particular.

I would say we might need to add more complexity around this cache so that agent keys stay in, even after the PassphraseManager locks itself up. But everything else gets cleared from the cache at some point, as a default that applies in particular to seeds.. Maybe dependent on the secret type? Or everything gets cleared except for `PRIMARY_KEYBUNDLE_ID`? What do you think?

Reply:
I guess I was imagining anything that needed persistence like for the app keys handling it on that end... i.e. keeping a copy of the secret after requesting it. But definitely something we can iterate on to see what works.